### PR TITLE
Feature(Scorecard): Added Metrics object & setting and updating of scorecard & base metrics.

### PR DESCRIPTION
**Pull request by szemyd**

### CODE REVIEW

File: src/krisi/evaluate/scorecard.py [hunk 1]
```diff
@@ -1,29 +1,209 @@
...
+            if isinstance(item, dict):
+                metric_dict = map_newdict_on_olddict(vars(metric), item)
+                self.__dict__[key] = Metric(**metric_dict)
+            elif isinstance(item, Metric):
+                metric_dict = map_newdict_on_olddict(vars(metric), vars(item))
+                self.__dict__[key] = Metric(**metric_dict)
+            else:
+                metric["metric_result"] = item
+                self.__dict__[key] = metric
+
```
Comment (szemyd) [inline src/krisi/evaluate/scorecard.py:130]: I created this method so we can do two things with scorecard:

1. Simply assign a value to a metric: 
`python
score_card.mse = 0.05
`
2. Add metadata along with the metric:
`python
score_cards.mse = Metric(0.05, hyperparameters={'alpha':0.05})
`
3. Update the hyperparameters as a secondary step if desired:
`python
score_card.mse = 0.05
score_card.mse.hyperparameters = {'alpha':0.05}

File: src/krisi/evaluate/evaluate.py [hunk 2]
```diff
@@ -2,21 +2,21 @@
...
 import numpy as np
 import pandas as pd
+from sklearn.metrics import mean_absolute_error, mean_squared_error
 from statsmodels.tsa.stattools import acf, pacf, q_stat
 
-from krisi.evaluate.scorecard import SampleTypes, ScoreCard
+from krisi.evaluate.scorecard import Metric, SampleTypes, ScoreCard
 
 
 def metric_hoc(func: Callable, *args, **kwargs) -> Callable:
```
Comment (itchingpixels) [inline src/krisi/evaluate/evaluate.py:11]: technically, this is called partial application of a function, not higher order component (that's React specific)

File: src/krisi/evaluate/scorecard.py [hunk 3]
```diff
@@ -1,29 +1,209 @@
...
+        return getattr(self, key, "Unknown Field")
+
+    def __str__(self) -> str:
+        return print_metric(self)
+
+    def __repr__(self) -> str:
+        return print_metric(self, repr=True)
+
+
+def map_newdict_on_olddict(old_dict: dict, new_dict: dict) -> dict:
```
Comment (itchingpixels) [inline src/krisi/evaluate/scorecard.py:48]: there are many ways to merge dictionaries, like the `|` operator https://favtutor.com/blogs/merge-dictionaries-python

File: src/krisi/evaluate/scorecard.py (hunk 1 above)
Comment (itchingpixels) [reply src/krisi/evaluate/scorecard.py:130]: is it useful to have the Scorecard to be mutable? why would we want to mutate it, rather than passing in all the variables at initalization? (if it's an internal thing)

File: src/krisi/evaluate/scorecard.py [hunk 4]
```diff
@@ -1,29 +1,209 @@
...
+    acf_res: Metric[Tuple[float, float]] = Metric(
+        name="Autocorrelation", category=MCats.residual
+    )
+    residuals_mean: Metric[float] = Metric(
+        name="Residual Mean", category=MCats.residual
+    )
+    residuals_std: Metric[float] = Metric(
+        name="Residual Standard Deviation", category=MCats.residual
+    )
+    # hearst_exponent: Metric[float] = Metric(category=MCats.entropy)
```
Comment (itchingpixels) [inline src/krisi/evaluate/scorecard.py:100]: hurst :)

File: src/krisi/evaluate/scorecard.py [hunk 5]
```diff
@@ -1,29 +1,209 @@
...
+@dataclass
+class Metric(Generic[T]):
+    name: str
+    category: Optional[MCats] = None
+    metric_result: Optional[T] = None
+    hyperparameters: Optional[Any] = None
+    info: str = ""
+
+    def __setitem__(self, key: str, item: Any) -> None:
+        setattr(self, key, item)
```
Comment (itchingpixels) [inline src/krisi/evaluate/scorecard.py:36]: why do we need __setitem__ here?

File: src/krisi/evaluate/scorecard.py [hunk 6]
```diff
@@ -1,29 +1,209 @@
...
+
+T = TypeVar("T", bound=Union[float, int, str, List, Tuple])
+
+
+@dataclass
+class Metric(Generic[T]):
+    name: str
+    category: Optional[MCats] = None
+    metric_result: Optional[T] = None
+    hyperparameters: Optional[Any] = None
```
Comment (itchingpixels) [inline src/krisi/evaluate/scorecard.py:32]: I think Metric can be immutable, so we can remove the setters and the @dataclass will automatically create an init function based on this.

Comment (itchingpixels) [reply src/krisi/evaluate/scorecard.py:32]: ahh okay, we can't, I see!

File: src/krisi/evaluate/scorecard.py [hunk 7]
```diff
@@ -1,29 +1,209 @@
...
+        info="Root Mean Squared Error",
+    )
+    aic: Metric[float] = Metric(
+        name="Akaike Information Criterion", category=MCats.entropy
+    )
+    bic: Metric[float] = Metric(
+        name="Bayesian Information Criterion", category=MCats.entropy
+    )
+    pacf_res: Metric[Tuple[float, float]] = Metric(
+        name="Partial Autocorrelation", category=MCats.residual
```
Comment (itchingpixels) [inline src/krisi/evaluate/scorecard.py:89]:
```suggestion
        name="Partial Autocorrelation of Residuals", category=MCats.residual
```

File: src/krisi/evaluate/scorecard.py [hunk 8]
```diff
@@ -1,29 +1,209 @@
...
+    )
+    acf_res: Metric[Tuple[float, float]] = Metric(
+        name="Autocorrelation", category=MCats.residual
```
Comment (itchingpixels) [inline src/krisi/evaluate/scorecard.py:92]:
```suggestion
        name="Autocorrelation of residuals", category=MCats.residual
```

File: src/krisi/evaluate/scorecard.py (hunk 1 above)
Comment (itchingpixels) [reply src/krisi/evaluate/scorecard.py:130]: I mean, something that we create ourselves, within this library

**Review by itchingpixels — APPROVED** [review]
👍

File: src/krisi/evaluate/evaluate.py (hunk 2 above)
Comment (szemyd) [reply src/krisi/evaluate/evaluate.py:11]: What would be a shorter word for it?

File: src/krisi/evaluate/scorecard.py (hunk 5 above)
Comment (szemyd) [reply src/krisi/evaluate/scorecard.py:36]: So we can use it as a dictionary!
